src/neo/Trie/MPT/MPTTrie.cs

```diff
@@ -0,0 +1,373 @@
+                            node = HashNode.EmptyNode();
+                            return true;
+                        }
+                        return false;
+                    }
+                case ShortNode shortNode:
+                    {
+                        var prefix = shortNode.Key.CommonPrefix(path);
+                        var oldHash = shortNode.GetHash();
+                        if (prefix.Length == shortNode.Key.Length)
+                        {
+                            var result = TryDelete(ref shortNode.Next, path.Skip(prefix.Length));
+                            if (!result) return false;
+                            db.Delete(oldHash);
+                            if (shortNode.Next is HashNode hashNode && hashNode.IsEmptyNode)
+                            {
+                                node = shortNode.Next;
+                            }
+                            else if (shortNode.Next is ShortNode sn)
+                            {
+                                shortNode.Key = shortNode.Key.Concat(sn.Key);
+                                shortNode.Next = sn.Next;
+                                shortNode.ResetFlag();
+                                db.Put(shortNode);
+                            }
+                            else
+                            {
+                                node.ResetFlag();
+                                db.Put(shortNode);
+                            }
+                            return true;
+                        }
+                        return false;
+                    }
+                case FullNode fullNode:
+                    {
+                        var result = false;
+                        var oldHash = fullNode.GetHash();
+                        if (path.Length == 0)
+                        {
+                            result = TryDelete(ref fullNode.Children[fullNode.Children.Length], path);
+                        }
+                        else
+                        {
+                            result = TryDelete(ref fullNode.Children[path[0]], path.Skip(1));
+                        }
+                        if (!result) return false;
+                        db.Delete(oldHash);
+                        var nonEmptyChildren = new byte[] { };
+                        for (int i = 0; i < fullNode.Children.Length; i++)
+                        {
+                            if (fullNode.Children[i] is HashNode hn && hn.IsEmptyNode) continue;
+                            nonEmptyChildren = nonEmptyChildren.Add((byte)i);
+                        }
+                        if (1 < nonEmptyChildren.Length)
+                        {
+                            fullNode.ResetFlag();
+                            db.Put(fullNode);
+                            return true;
+                        }
+                        var childIndex = nonEmptyChildren[0];
```

Comment:
Is it possible the size of nonEmptyChildren can be zero?

Reply:
No. If there is, there must be a bug.
A FullNode should have at least 2 children. If not, we use ShortNode instead. We should check thist after every Delete.